R/rules.R

```diff
@@ -169,121 +177,150 @@ sv_regex <- function(pattern, message, ignore.case = FALSE, perl = FALSE,
-#'   iv$add_rule("count", ~if (. <= 0) "A positive value is required")
 #'
+#'   iv$add_rule("count", sv_numeric())
+#'   iv$add_rule("count", ~if (. <= 0) "A positive value is required")
 #' })
+#'
 #' @export
-sv_numeric <- function(message = "A number is required", allowMultiple = FALSE,
-  allowNA = FALSE, allowNaN = FALSE, allowInfinite = FALSE) {
-
+sv_numeric <- function(message = "A number is required",
+                       allow_multiple = FALSE,
+                       allow_na = FALSE,
+                       allow_nan = FALSE,
+                       allow_inf = FALSE) {
   force(message)
-  force(allowMultiple)
-  force(allowNA)
-  force(allowNaN)
-  force(allowInfinite)
+  force(allow_multiple)
+  force(allow_na)
+  force(allow_nan)
+  force(allow_inf)
 
-  function(value) {
-    if (!is.numeric(value)) {
-      return(message)
-    }
-    if (length(value) == 0) {
-      return(message)
-    }
-    if (!allowMultiple && length(value) != 1) {
-      return(message)
-    }
-    if (!allowNA && any(is.na(value))) {
-      return(message)
-    }
-    if (!allowNaN && any(is.nan(value))) {
-      return(message)
-    }
-    if (!allowInfinite && any(!is.finite(value))) {
-      return(message)
+  compose_rules(
+    sv_basic(
+      allow_multiple = allow_multiple,
+      allow_na = allow_na,
+      allow_nan = allow_nan,
+      allow_inf = allow_inf
+    ),
+    function(value) {
+      
+      if (length(value) == 0) {
+        return(err_condition_messages[["err_zero_length_value"]])
+      }
+      
+      # Validation test
+      res <- is.numeric(value) || is.infinite(value) || is.na(value)
+      
+      if (!all(res)) {
+        return(message)
+      }
     }
-  }
+  )
 }
 
-#' @rdname sv_numeric
+#' Validate that a field is an integer
+#'
+#' The `sv_integer()` function validates that a field is an integer with the
+#' [base::is.integer()] function. By default, only a single, finite,
+#' not-missing, valid integer is allowed, but each of those criteria can be
+#' controlled via arguments.
+#'
+#' @param message The validation error message to use if a value is not an
+#'   integer.
+#' @inheritParams sv_numeric
+#' 
+#' @return A function suitable for using as an
+#'   [`InputValidator$add_rule()`][InputValidator] rule.
+#'
+#' @examples
+#' # Ignore withReactiveDomain(), it's just required to get this example to run
+#' # outside of Shiny
+#' shiny::withReactiveDomain(shiny::MockShinySession$new(), {
+#'
+#'   iv <- InputValidator$new()
+#'
+#'   iv$add_rule("count", sv_integer())
+#'   iv$add_rule("count", ~if (. <= 0) "A positive value is required")
+#' })
+#'
 #' @export
 sv_integer <- function(message = "An integer is required",
-  allowMultiple = FALSE, allowNA = FALSE, allowNaN = FALSE,
-  allowInfinite = FALSE) {
-  
+                       allow_multiple = FALSE,
+                       allow_na = FALSE,
+                       allow_nan = FALSE,
+                       allow_inf = FALSE) {
```

Comment:
Infinity can't be an integer, so remove this param and always pass `allow_inf = FALSE` to `sv_basic`

Reply:
Now done. Checking for integer-like values with `value %% 1 == 0` is seemingly no better and no worse than `rlang::is_integerish()`. Added a note in the documentation that very large integers (~ >1E15 won't) probably validate correctly (for reasonably large floating point values like 0.1).

Used `suppressWarnings(value %% 1 == 0)` to avoid an R warning about the `%%` losing accuracy for larger LHS values.

For larger integer-like numbers, we might be able to come up with a better solution in the future.